contracts/staking/StakeHolder.sol

```diff
@@ -0,0 +1,266 @@
+    ///    a balance of 0). This array is never re-ordered. As such, off-chain services
+    ///    could cache the results of getStakers().
+    // solhint-disable-next-line private-vars-leading-underscore
+    address[] private stakers;
+
+    /// @notice version number of the storage variable layout.
+    uint256 public version;
+
+    /**
+     * @notice Initialises the upgradeable contract, setting up admin accounts.
+     * @param _roleAdmin the address to grant `DEFAULT_ADMIN_ROLE` to
+     * @param _upgradeAdmin the address to grant `UPGRADE_ROLE` to
+     */
+    function initialize(address _roleAdmin, address _upgradeAdmin) public initializer {
+        __UUPSUpgradeable_init();
+        __AccessControl_init();
+        _grantRole(DEFAULT_ADMIN_ROLE, _roleAdmin);
+        _grantRole(UPGRADE_ROLE, _upgradeAdmin);
+        version = _VERSION0;
+    }
+
+    /**
+     * @notice Function to be called when upgrading this contract.
+     * @dev Call this function as part of upgradeToAndCall().
+     *      This initial version of this function reverts. There is no situation
+     *      in which it makes sense to upgrade to the V0 storage layout.
+     *      Note that this function is permissionless. Future versions must
+     *      compare the code version and the storage version and upgrade
+     *      appropriately. As such, the code will revert if an attacker calls
+     *      this function attempting a malicious upgrade.
+     * @ param _data ABI encoded data to be used as part of the contract storage upgrade.
+     */
+    function upgradeStorage(bytes memory /* _data */) external virtual {
+        revert CanNotUpgradeToLowerOrSameVersion(version);
+    }
+
+    /**
+     * @notice Allow any account to stake more value.
+     * @dev The amount being staked is the value of msg.value.
+     * @dev This function does not need re-entrancy guard as the add stake
+     *  mechanism does not call out to any external function.
+     */
+    function stake() external payable {
+        if (msg.value == 0) {
+            revert MustStakeMoreThanZero();
+        }
+        _addStake(msg.sender, msg.value, false);
+    }
+
+    /**
+     * @notice Allow any account to remove some or all of their own stake.
+     * @dev This function does not need re-entrancy guard as the state is updated
+     *  prior to the call to the user's wallet.
+     * @param _amountToUnstake Amount of stake to remove.
+     */
+    function unstake(uint256 _amountToUnstake) external {
+        StakeInfo storage stakeInfo = balances[msg.sender];
+        uint256 currentStake = stakeInfo.stake;
+        if (currentStake < _amountToUnstake) {
+            revert UnstakeAmountExceedsBalance(_amountToUnstake, currentStake);
+        }
+        uint256 newBalance = currentStake - _amountToUnstake;
+        stakeInfo.stake = newBalance;
+
+        emit StakeRemoved(msg.sender, _amountToUnstake, newBalance);
```

Comment:
out of curiosity, why emit first and then transfer? or it doesn't matter?

Reply:
I moved this emit prior to the transfer as a result of a warning from slither. My guess at why this is deemed better is that there would be two StakeRemoved events emitted if the transfer caused the code to re-enter the function, which would make it easier to detect.
However, there is a test for re-entrancy, and given transfer only supplies 3000 gas, even a minimal re-entrancy attack contract is unable to successfully execute an attack.